R/parse.R

```diff
@@ -119,19 +119,106 @@ parse_check_consistent_dimensions <- function(dat, call) {
     if (!is.null(eq$lhs)) {
       parse_check_consistent_dimensions_lhs(eq, dat, call)
     }
+    if (!is.null(eq$rhs)) {
+      parse_check_consistent_dimensions_rhs(eq, dat, call)
+    }
   }
 }
 
-parse_check_consistent_dimensions_lhs <- function(eq, dat, call) {
-  if (!is.null(eq$lhs$array)) {
-    rank <- length(eq$lhs$array)
-    dim_rank <- dat$storage$arrays$rank[dat$storage$arrays$name == eq$lhs$name]
-    if (rank != dim_rank) {
-      odin_parse_error(
-        c("Array rank in expression differs from the rank declared with `dim`",
-        i = paste("'{eq$lhs$name}' has rank '{dim_rank}' in the `dim` call, ",
-                  "but the line below assumes it has rank '{rank}'.")),
-        "E2018", eq$src, call)
+parse_check_consistent_dimensions_lhs <- function(eq, dat, call, src = eq$src) {
+  throw_mismatch <- function(var, dim_rank, array_rank) {
+    odin_parse_error(
+      c("Array rank in expression differs from the rank declared with `dim`",
+        i = paste("'{var}' has rank '{dim_rank}' in the `dim` call, ",
+                  "but the line below assumes it has rank '{array_rank}'.")),
+      "E2018", src, call)
+  }
```

Comment:
This is duplicated from rhs - but I didn't really like it being either a level higher, or merging the lhs and rhs functions under one parent...